keps/1337-reclaim-within-cohort-while-borrowing/README.md

```diff
@@ -0,0 +1,373 @@
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- support preemption to reclaim quota within cohort for workloads that require borrowing.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- allow preemption of workloads, from other ClusterQueues, which are not borrowing.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Introduce a new ClusterLevel API to support preemption while borrowing. The
+API proposes additional configuration to the `.preemption.reclaimWithinCohort`
+option.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As an administrator of Kueue in my organization I setup the following configuration:
+there are multiple cluster queues in a cohort, corresponding to different teams.
+All of the cluster queues have `nominalQuota: 0`, and `borrowingLimit>0`. In
+this cohort, there is also a single shared cluster queue with `nominalQuota>0`,
+and `borrowingLimit=0`. In this setup every workload requires borrowing. In this
+setup I would like to enable preemption for high priority workloads.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Flopping between preempting pair of workloads
+
+In a setup as described in (Story 1)[#story-1] consider there are two workloads,
+one in ClusterQueue A, and one in ClusterQueue B. There is not enough resources
+in the cohort to run both workloads at the same time.
+
+If the preemption is allowed for workloads of any priorities
+(as when `.preemption.reclaimWithinCohort` equals `Any`)
+then the following flopping happens: workload A preempts workload B, then p
+preempted workload B returns to the queue and preempts workload A to make progress.
+This repeats, and none of the workloads can make any progress.
```

Comment:
Yeah, good point. For example, when `.preemption.withinClusterQueue: Any` - we offer `Any`. I think it might an argument for `Policy` enum, rather than just `Enable/Disable`, because maybe in the future we want to offer `Any` in this case also.

Reply:
Added a note in alternatives, that Policy might be needed in the future for the reason.

Reply:
Actually no, this `.preemption.withinClusterQueue: Any` isn't supported currently, probably for this exact reason. In that case, it's not clear to me this flopping can currently happen. We want to design the API so that users don't run into it.